service/worker/dlq/workflow.go

```diff
@@ -0,0 +1,161 @@
+	// WorkflowParams is the single argument to the DLQ workflow.
+	WorkflowParams struct {
+		// WorkflowType options are available via the WorkflowType* constants.
+		WorkflowType string
+		// DeleteParams is only used for WorkflowTypeDelete.
+		DeleteParams DeleteParams
+	}
+	// DeleteParams contain the target DLQ and the max message ID to delete up to.
+	DeleteParams struct {
+		TaskCategory  int
+		SourceCluster string
+		TargetCluster string
+		MaxMessageID  int64
+	}
+	// HistoryServiceClient is a subset of the [historyservice.HistoryServiceClient] interface.
+	HistoryServiceClient interface {
+		DeleteDLQTasks(
+			ctx context.Context,
+			in *historyservice.DeleteDLQTasksRequest,
+			opts ...grpc.CallOption,
+		) (*historyservice.DeleteDLQTasksResponse, error)
+	}
+	workerComponent struct {
+		client HistoryServiceClient
+	}
+)
+
+const (
+	// WorkflowName is the name of the DLQ workflow.
+	WorkflowName = "temporal-sys-dlq-workflow"
+	// WorkflowTypeDelete is what the value of WorkflowParams.WorkflowType should be to delete DLQ tasks.
+	WorkflowTypeDelete = "delete"
+
+	errorTypeInvalidWorkflowType = "dlq-invalid-workflow-type"
+	errorTypeUnavailable         = "history-service-unavailable"
+	errorTypeUnknown             = "unknown-error"
+	deleteTasksActivityName      = "dlq-delete-tasks-activity"
+
+	deleteTasksTimeout = 15 * time.Second
+)
+
+// Module provides a [workercommon.WorkerComponent] annotated with [workercommon.WorkerComponentTag] to the graph, given
+// a [HistoryServiceClient] dependency.
+var Module = workercommon.AnnotateWorkerComponentProvider(newComponent)
+
+func newComponent(client HistoryServiceClient) workercommon.WorkerComponent {
+	return &workerComponent{
+		client: client,
+	}
+}
+
+//revive:disable:import-shadowing this doesn't actually shadow imports because it's a method, not a function
+func (c *workerComponent) workflow(ctx workflow.Context, params WorkflowParams) error {
+	if params.WorkflowType == WorkflowTypeDelete {
+		ctx = workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
+			StartToCloseTimeout: deleteTasksTimeout,
+		})
+		future := workflow.ExecuteActivity(ctx, deleteTasksActivityName, params.DeleteParams)
+		return future.Get(ctx, nil)
+	}
+	return temporal.NewNonRetryableApplicationError(params.WorkflowType, errorTypeInvalidWorkflowType, nil)
+}
+
+func (c *workerComponent) deleteTasks(ctx context.Context, params DeleteParams) error {
+	req := &historyservice.DeleteDLQTasksRequest{
+		DlqKey: &commonspb.HistoryDLQKey{
+			TaskCategory:  int32(params.TaskCategory),
+			SourceCluster: params.SourceCluster,
+			TargetCluster: params.TargetCluster,
+		},
+		InclusiveMaxTaskMetadata: &commonspb.HistoryDLQTaskMetadata{
+			MessageId: params.MaxMessageID,
+		},
+	}
+	_, err := c.client.DeleteDLQTasks(ctx, req)
+	if err != nil {
+		var svcErr serviceerror.ServiceError
+		// Use this instead of serviceerror.ToStatus because that doesn't work for wrapped errors
+		if errors.As(err, &svcErr) {
```

Comment:
is context timeout a ServiceError?

Reply:
Changed this logic so that everything is retryable except not-found and invalid argument errors.